corpus_studies/solidity_state_study/slither_plugins/EnumExplorer.py

```diff
@@ -0,0 +1,205 @@
+from slither.core.expressions import *
+from slither.core.cfg.node import *
+from slither.core.declarations import *
+from slither.detectors.functions.modifier import is_revert
+from .Graph import Edge, TransitionGraph
+
+# Checks if the expression is an enum constant of the type enum_type.
+def expr_is_enum_val(exp: Expression, enum_type: EnumContract) -> bool:
+    return isinstance(exp, MemberAccess) and isinstance(exp.expression, Identifier) and \
+        exp.expression.value == enum_type
+
+# Checks if the expression is equal to the variable var.
+def expr_is_var(exp: Expression, var: Variable) -> bool:
+    return isinstance(exp, Identifier) and exp.value == var
+
+class EnumExplorer:
+    def __init__(self, 
+                contract: Contract,
+                state_var: StateVariable, 
+                enum_type: EnumContract):
+        self.contract = contract
+        self.state_var = state_var
+        self.all_states = set(enum_type.values)
+        self.enum_type = enum_type
+
+    def inferTransitionGraph(self) -> TransitionGraph:
+        # Find the variable constructor, if it exists (where fields are initialized when they are declared).
+        constructor_vars = next((f for f in self.contract.functions if f.is_constructor_variables), None)
+        # Find the contract's constructor, if it exists.
+        # A contract can have a most one constructor, not counting inherited constructors.
+        # For a contract C, we are looking for a function with canonical name "C.constructor()"
+        # or "C.C()", depending on the version of Solidity.
+        constructor = next((f for f in self.contract.functions if f.is_constructor and f.canonical_name.startswith(self.contract.name)), None)
+        initial_states = self.find_initial_states(constructor_vars, constructor)
+
+        graph = TransitionGraph(set(self.enum_type.values), initial_states)
+
+        for func in self.contract.functions:
+            if not (func.is_constructor or func.is_constructor_variables):
+                edges = self.exploreFunction(func)
+                for edge in edges:
+                    graph.addEdge(edge.s_from, edge.s_to, edge.label)
+        return graph
+
+    def find_initial_states(self, constructor_vars, constructor) -> Set[str]:
+        initial_states = {self.enum_type.values[0]}
+        if constructor_vars is not None:
+            edges = self.exploreFunction(constructor_vars)
+            initial_states = {edge.s_to for edge in edges if edge.s_from in initial_states}
+        if constructor is not None:
+            edges = self.exploreFunction(constructor)
+            initial_states = {edge.s_to for edge in edges if edge.s_from in initial_states}
+        return initial_states
+
+    def exploreFunction(self, func: Function) -> List[Edge]:
```

Comment:
As with the filename, "explore" is not a specific enough name here. Say what the function actually does.

Reply:
And likewise for other similarly-named functions (e.g. exploreNode).